tests/frontend/test_tdmprogram.py

```diff
@@ -0,0 +1,425 @@
+    concurrA = 1 + delayA
+    concurrB = 1 + delayB
+    concurrC = 1 + delayC
+    concurrD = 1 + delayD
+
+    N = [concurrA, concurrB, concurrC, concurrD]
+
+    sq_r = 5
+
+    # first half of cluster state measured in X, second half in P
+    n = 400  # number of timebins
+    theta_A = [0] * int(n / 2) + [np.pi / 2] * int(n / 2)  # measurement angles for detector A
+    theta_B = theta_A  # measurement angles for detector B
+    theta_C = theta_A
+    theta_D = theta_A
+
+    # 2D cluster
+    prog = tdmprogram.TDMProgram(N)
+    with prog.context(theta_A, theta_B, theta_C, theta_D, shift="end") as (p, q):
+
+        ops.Sgate(sq_r, 0) | q[0]
+        ops.Sgate(sq_r, 0) | q[2]
+        ops.Sgate(sq_r, 0) | q[8]
+        ops.Sgate(sq_r, 0) | q[9]
+
+        ops.Rgate(np.pi / 2) | q[0]
+        ops.Rgate(np.pi / 2) | q[8]
+
+        ops.BSgate(np.pi / 4) | (q[0], q[2])
+        ops.BSgate(np.pi / 4) | (q[8], q[9])
+        ops.BSgate(np.pi / 4) | (q[2], q[8])
+        ops.BSgate(np.pi / 4) | (q[0], q[1])
+        ops.BSgate(np.pi / 4) | (q[3], q[9])
+
+        ops.MeasureHomodyne(p[0]) | q[0]
+        ops.MeasureHomodyne(p[1]) | q[1]
+        ops.MeasureHomodyne(p[2]) | q[3]
+        ops.MeasureHomodyne(p[3]) | q[9]
+
+    eng = sf.Engine("gaussian")
+    result = eng.run(prog)
+    samples = result.all_samples
+
+    xA = result.all_samples[0]
+    xB = result.all_samples[1]
+    xC = result.all_samples[3]
+    xD = result.all_samples[9]
+
+    X_A = xA[: n // 2]  # X samples from detector A
+    P_A = xA[n // 2 :]  # P samples from detector A
+
+    X_B = xB[: n // 2]  # X samples from detector B
+    P_B = xB[n // 2 :]  # P samples from detector B
+
+    X_C = xC[: n // 2]  # X samples from detector C
+    P_C = xC[n // 2 :]  # P samples from detector C
+
+    X_D = xD[: n // 2]  # X samples from detector D
+    P_D = xD[n // 2 :]  # P samples from detector D
+
+    N = delayC
+    # nullifiers defined in https://arxiv.org/pdf/1903.03918.pdf, Fig. S5
+    ntot = len(X_A) - N - 1
+    nX1 = np.array([X_A[k] + X_B[k] - np.sqrt(1 / 2) * (-X_A[k + 1] + X_B[k + 1] + X_C[k + N] + X_D[k + N]) for k in range(ntot)])
+    nX2 = np.array([X_C[k] - X_D[k] - np.sqrt(1 / 2) * (-X_A[k + 1] + X_B[k + 1] - X_C[k + N] - X_D[k + N]) for k in range(ntot)])
+    nP1 = np.array([P_A[k] + P_B[k] + np.sqrt(1 / 2) * (-P_A[k + 1] + P_B[k + 1] + P_C[k + N] + P_D[k + N]) for k in range(ntot)])
+    nP2 = np.array([P_C[k] - P_D[k] + np.sqrt(1 / 2) * (-P_A[k + 1] + P_B[k + 1] - P_C[k + N] - P_D[k + N]) for k in range(ntot)])
+
+    nX1var = np.var(nX1)
+    nX2var = np.var(nX2)
+    nP1var = np.var(nP1)
+    nP2var = np.var(nP2)
+
+    assert np.allclose(nX1var, 4 * np.exp(-2 * sq_r), rtol=5 / np.sqrt(ntot))
+    assert np.allclose(nX2var, 4 * np.exp(-2 * sq_r), rtol=5 / np.sqrt(ntot))
+    assert np.allclose(nP1var, 4 * np.exp(-2 * sq_r), rtol=5 / np.sqrt(ntot))
+    assert np.allclose(nP2var, 4 * np.exp(-2 * sq_r), rtol=5 / np.sqrt(ntot))
```

Comment:
Thanks for formalizing all these tests, it looks great! One day, you'll have to teach me about `assert` and `allclose`.

Reply:
when running `pytest tests/frontend/test_tdmprogram.py -x --tb=short --cov=strawberryfields/tdm --cov-report term-missing` I get
![image](https://user-images.githubusercontent.com/991946/92187710-8672e280-ee28-11ea-9e56-1979efb6912d.png)

Reply:
The tests look great @fab1an-q and @nquesada!